crates/boundless-market/src/contracts/mod.rs

```diff
@@ -553,20 +567,249 @@ impl Requirements {
+                    Digest::try_from(image_id).map_err(|_| PredicateError::DataMalformed)?,
+                    journal_prefix.to_vec().into(),
+                ))
+            }
+            PredicateType::ClaimDigestMatch => {
+                // Don't need to check the length because that is checked when converting to Digest
+                Ok(Predicate::ClaimDigestMatch(
+                    Digest::try_from(value.data.as_ref())
+                        .map_err(|_| PredicateError::DataMalformed)?,
+                ))
+            }
+            PredicateType::__Invalid => Err(PredicateError::InvalidType),
+        }
+    }
+}
+
 impl Predicate {
     /// Returns a predicate to match the journal digest. This ensures that the request's
     /// fulfillment will contain a journal with the same digest.
-    pub fn digest_match(digest: impl Into<Digest>) -> Self {
-        Self {
-            predicateType: PredicateType::DigestMatch,
-            data: digest.into().as_bytes().to_vec().into(),
-        }
+    pub fn digest_match(image_id: impl Into<Digest>, journal_digest: impl Into<Digest>) -> Self {
+        Self::DigestMatch(image_id.into(), journal_digest.into())
     }
-
     /// Returns a predicate to match the journal prefix. This ensures that the request's
     /// fulfillment will contain a journal with the same prefix.
-    pub fn prefix_match(prefix: impl Into<Bytes>) -> Self {
-        Self { predicateType: PredicateType::PrefixMatch, data: prefix.into() }
+    pub fn prefix_match(image_id: impl Into<Digest>, journal_prefix: impl Into<Bytes>) -> Self {
+        Self::PrefixMatch(image_id.into(), journal_prefix.into())
+    }
+    /// Returns a predicate to match the claim digest. This ensures that the request's
+    /// fulfillment will contain a claim with the same digest.
+    pub fn claim_digest_match(claim_digest: impl Into<Digest>) -> Self {
+        Self::ClaimDigestMatch(claim_digest.into())
+    }
+
+    /// Returns the image_id if it exists.
+    pub fn image_id(&self) -> Option<Digest> {
+        match self {
+            Predicate::DigestMatch(image_id, ..) => Some(*image_id),
+            Predicate::PrefixMatch(image_id, ..) => Some(*image_id),
+            Predicate::ClaimDigestMatch(..) => None,
+        }
+    }
+
+    /// Evaluates the predicate against the fulfillment data, returning the claim digest if the
+    /// evaluation succeeds or `None` if it fails.
+    pub fn eval(&self, fulfillment_data: &FulfillmentData) -> Option<Digest> {
+        let claim_digest_data = match fulfillment_data {
+            FulfillmentData::None => None,
+            FulfillmentData::ImageIdAndJournal(image_id, journal) => {
+                Some(ReceiptClaim::ok(*image_id, journal.to_vec()).digest())
+            }
+        };
+        let claim_digest_predicate = match self {
+            Predicate::DigestMatch(image_id, journal) => {
+                Some(ReceiptClaim::ok(*image_id, MaybePruned::Pruned(*journal)).digest())
```

Comment:
I missed a check when I wrote this, to require the prover to deliver the journal.
```suggestion
                // With the DigestMatch predicate, we require the delivery of the journal.
                // If is checked to match the predicate by the claim digest comparison below.
                let FulfillmentData::ImageIdAndJournal(_, _) = &fulfillment_data else {
                    return None;
                };
                Some(ReceiptClaim::ok(*image_id, MaybePruned::Pruned(*journal)).digest())
```